### ISSUE TITLE
ENH: support multiple detectors in same scan

### ISSUE THREAD

**Issue by CJ-Wright** [xpdAcq/mission-control#54]
## Expected Behavior
Experiment might be using 2 detectors (near-field, far-field).
We want to run_calibration only once for each detector.
When users switch detectors, software knows what calibration file should be used.

## Current Behavior
We only cache the calibration information for the most recent calibration.
 
## Possible Solution
Cache the calibration info somewhere then put it into the start depending on which detector is being used.
This may be an aspect of beamline hashing

**Comment by sbillinge** [comment]
Question: When the BLS change detectors, do they give the detector a
different PV?  This would immediately solve our problem because we could
link a calibration to the PV of the detector that was calibrated.

If (as I have a vague recollection is the case) the detector cabling gets
physically switched and comes through the same PV, then perhaps we should
discuss about how to do it with different PVs which seems like a more
robust fix.  This may require a hardware upgrade from the BLS's so let's
discuss.

This would also allow for the UC that we run both detectors in the same
experiment, which is something I am interesting in exploring....i.e., beam
center is slightly offset from the side of the detector (or goes through a
hole in the detector) and a second detector is running behind.

S

**Comment by CJ-Wright** [comment]
The detectors are currently serviced by the same interface card, so the take on the same PV.

Multiple cards would make this easier. Does anyone at @xpdAcq/bls know what the status of this is?

I imagine that we could actually have the beam hit the SAXS detector and reconstruct the full pattern by stitching the WAXS and SAXS I(Q)s together.

**Comment by sbillinge** [comment]
Suggest we table this till we have different PVs per detector. Workaround
is to run a new calibration each time detector is switched. Then by default
the subsequent runs should pick the right calibration since it is the most
recent.

**Comment by CJ-Wright** [comment]
I guess, although if every data needed saxs and waxs data then either the user needs to load the sample twice or calibrate twice per sample.

This may be a good way to get the saxs or waxs metadata into the database.

We can table this, but I don't think think that having separate PVs will help us much. We will still need to build something that a) caches the calibrations for each detector, b) understands which calibration goes with which detector, c) understands when each calibration is no longer applicable and d) knows which detector is being used (and which calibration to use).  In the case of not having the PVs we could have the users track this for us by switching something in the globals, in the case with PVs this would be handled automatically by examining which detector is being used.

**Comment by sghose** [comment]
Guys,
This is what called a brainstorming session. The very purpose of using two PE detectors are the followings:
1. We are working on making the beamline a bit automated operation switching between different detector configurations and temperature configurations.
2. Later we will add the Robots sample changes with a new sample type we have commissioned.
3. In this process I am working making combined SAXS/PDF and XRD/PDF measurements protocols.

All the plans are in the same line of future projects that we envisioned for XPD/PDF for data collection and Analysis. We all are working to getting to the goal one step at a time. There are many technical developmentals planning steps to be implemented and commissioned for science before we integrate everything. One of the small step is to positioning/motion of two detectors and device the xpdAcq/xpdAn protocol for combined operation of these detectors. This is what the current discussion will focused on. The measurement protocol depends up on what capabilities we have and how optize the scheme to address the science we need to follow.
 Detector Configuration and Motion: (Status):
1. We will be running two detectors at XPD as before one called PE1 ( PDF) other one PE2( far field, XRD/SAXS).
2. Both the detectors will run parallely and spit out two PVs. How we do it is up to XPD staffs. Just to mention, we have two cards and I tested them long back (1.5 yrs) to run both detector parallel. We will do the same but with two PVs. We will later ( very soon) will replace PE1 with a new PE detector with it's own computer. 
3. Both the detectors will have X & Z motions ( the PE1 will have a fast X ( 200mm/sec). 
4. Status: so far fast motion is not ready, neither parallel detector implementation is ready. The parallel operation will be ready in a week time. So, need not to worry on this on 16th/17th upgrade.

We need the following things to be implemented when both the detectors are ready for parallel operation.
1. Detectors are on all the time as before ( but with twp PVs). We collect data on the one we asked for. The data will be saved in the same way but PE1 data will go to pe1_data and PE2 data will go to pe2_data folders.
2. Need to work out the Integration & Analysis procedure.

Cheers!!!!

**Comment by dooryhee** [comment]
PE1 (2k) and PE2 (4k) are ready for use with xpdAcq. PE3 (2k) in the works. We're planning on operating PE1 and PE2 simultaneously in 2018-2 cycle (start date  May 29).

**Comment by CJ-Wright** [comment]
I think the main issue here is one of user interface. On the internal side we will associate each detector with it's own descriptor/event stream. On the analysis side we will process the data separately so there should be no conflicts.

On the user side this is more tricky. Currently we only support one detector, which is essentially hard-wired inside `xrun`. This makes life for the users easier, as they don't need to specify the detector. Now we will need to specify detector parameters (eg. how long to take shots) on a per detector basis. Additionally we will need to add logic to all the shots to move the near detector so we can get a clear shot.

How commonly do you expect to run the two detectors in the same measurement?

**Comment by sghose** [comment]
I think we should first have all the detectors run through xpdAcq as per they are called. Simultaneous operation could be done at a later stage.

I would go for a flexibility in measurement and analysis as we chose the detectors. The user will specify what they need to run.
We could limit this to PDF(Near) PE1 or XRD/SAXS(far) PE2/PE3. This way we could accommodate any detectors included to XPD acquisition.
The analysis will go with the detector type and it’s output size. Because, in future we will have a dexela which size is different.

Just to update you about the data storage:

1.       We will have now the network storage partitioned by two ( XPD and PDF). The servers serves the clients through GPFS ( which is faster with multiple channel queries)

2.       At XPD we will be having folders pe1_data, pe2_data, pe3_data, … as per the detector tiff file storage. EPICS configure this by default.

3.       The filestore will point to these folders for respective data.

4.       This needs discussion further: My suggestion would be when we choose the directory all the acquisition files, xpdUser folder created as per the detector you chose.

5.       At the end beamtime it should also archive in the same folder.

**Comment by sbillinge** [comment]
Hi Everyone,

We need a user UC to code to.  How about this:
1. user wants to interleave SAXS and WAXS scans during their experiment
2. instrument scientist aligns the beamstop on the SAXS detector
3. user determines the optimal data acquisition conditions for the SAXS detector
4. IS aligns the beamstop on the WAXS detector
5. user determines the optimal data acquisition conditions for the WAXS detector
6. user requests a `SAXS-WAXSinterleaf` scan where at each time point measurements are made on each detector.
7. software drives the WAXS detector in and takes the first point under optimal conditions
8. software drives the WAXS detector out
9. software exposes the SAXS detector using the optimal settings
10. software resets so the process can be repeated. 
11. WAXS stream is sent into xpdAn regular PDF pipeline
12. SAXS stream is sent into a similar pipeline that visualizes the 2D data in the same way as the WAXS pipeline plus an I(Q) on a linear and an I(Q) on a log scale
13. Data-files are dumped to disc in the normal way.

**Comment by sbillinge** [comment]
@sghose @CJ-Wright @chiahaoliu I suggest that we move this off the cycle 2 milestone.  First I would like to be sure that we need joint SAXS WAXS in the initial experiments in June, and second, even if we do, we could try and deliver this for June rather than May and have a mini-release mid-cycle

**Comment by dooryhee** [comment]
-        calibration info needs to be saved and interchanged automatically

-        this WE, users were running 90+ samples with both the PDF and XRD detectors. Each sample was specific, meaning statistics and saturation varied a lot and unpredictably. It makes me realize that automation should also enable software to determine saturation (saturation happening at Bragg spots or over the entire diffraction rings), minimize ghosting, and keep statistics at its best (points 3 and 5 below). Without that, users have to check all data individually, meaning that robot ops loses a lot of value.

**Comment by CJ-Wright** [comment]
TODO (not complete listing):
- [ ] Calibration needs to be stored in the graph independently (this will be done by having two separate graphs)
- [ ] Calibration needs to separated when stuffed into data, ether by having it in a dict with keys that relate to the detectors or in the descriptors?
- [ ] SAXS and WAXS need to be on different descriptors

**Comment by CJ-Wright** [comment]
Update on architecture:
Talking with @chiahaoliu I think we have come up with a good way going forward. We can use the `configure` plan stub to update the detector configuration with the calibration information.

Sequence for XPD:
1. User takes first calibration data on two detectors
1. xpdAn writes the output calibration information for each detector to a file `<detector_name>_calib.yaml`
1. User runs next run
1. xpdAcq reads the calibration data from the file (plus any motor readbacks that we captured during the calibration process) and configures the detector to hold this information. Or we hold an external cache of data and update the configuration of the detectors based off of the keys in the cache (time, motor positions)
1. The run starts
1. A descriptor for each detector being used comes down including that detector's calibration (maybe the entire cache?)
1. The run continues
1. The analysis software uses the calibration in the descriptor to run the analysis.

Sequence for PDF:
Similar to XPD but instead of two detectors we report the calibration at that motor position.

It would be nice to be able to update the detector configuration without emitting a new descriptor (since trigger and read will do that for us).

**Comment by dooryhee** [comment]
We’re aiming for the hardware to be in place this Summer during shutdown. Would be good if the below workflow could be tested before 2018-3 start-up. Thanks. E

**Comment by CJ-Wright** [comment]
Note to self (from @danielballan)
```python
class MyDetector(SomeOphydThing):
    def __init__(self, *args, calibration, **kwargs):
        super().__init__(*args, **kwargs)
        self.calibration = calibration
    def describe(self):
        ret = super().describe()
        ret['calibration'] = calibration

calibration = {...}  # hard-coded or use zict
my_detector = MyDetector(calibration=calibration)
```

**Comment by CJ-Wright** [comment]
The detectors also need to describe their pixel size for pyFAI

**Comment by CJ-Wright** [comment]
This might also be pending the ``Router`` class for pipeline management

**Comment by tacaswell** [comment]
I would rather see this done with `Device` classes than by mutating the descriptor as it goes by.  Something like

```
class CalibrationData(Device):
     tilt1 = Cpt(Signal, value=0, egu='radian', kind='config')
     tilt2 = Cpt(Signal, value=0, egu='radian', kind='config')
     dist = Cpt(Signal, value=1, egu='m', kind='config')

class DET(AD):
    calib = Cpt(Calibration, kind='config')
```

and then on stage have the logic to read the calibration from disk / what ever and update the PVs or host the values in a caproto IOC.

**Comment by sghose** [comment]
Hi All,
    I would like this issue to be reopened and to be brought to highest priority as required for the 28ID beamlines. I would request Tom Caswel to pick up this issue and deploy the work management how to proceed on this through DAMA group. I had a quick discussion with Max (DAMA group) in presence of Tim and Max mentioned this would be possible with the new BS release next. 
All I need to run two detectors one after the other with their inline data analysis/visualization/saving with different calibrations. Later we might extend this to use both the detector simultaneously. Testing can be done during cycle 2020-2 in May.
Thanks,
Sanjit

**Comment by st3107** [comment]
Target for 2021-2 cycle

**Comment by st3107** [comment]
I overhauled the xpdacq underlying infrasturcutre. It is working for multiple moving detectors to automatically find the calibration and dark.

I think the dark protocol still needs some improvement, but it is working overall.

**Comment by st3107** [comment]
Clove the issue because of the delivered features. Maybe a new issue of improvement will be done in the future.
